state/azure/tablestorage/tablestorage.go

```diff
@@ -187,34 +200,37 @@ func getTablesMetadata(metadata map[string]string) (*tablesMetadata, error) {
 		return nil, errors.New(fmt.Sprintf("missing or empty %s field from metadata", tableNameKey))
 	}
 
+	if val, ok := metadata[cosmosDBmodeKey]; ok && val != "" {
+		meta.cosmosDBmode = val == "true"
+	} else {
+		meta.cosmosDBmode = false
+	}
+
 	return &meta, nil
 }
 
 func (r *StateStore) writeRow(req *state.SetRequest) error {
-	pk, rk := getPartitionAndRowKey(req.Key)
-	entity := r.table.GetEntityReference(pk, rk)
-	entity.Properties = map[string]interface{}{
-		valueEntityProperty: r.marshal(req),
-	}
-
-	var etag string
-	if req.ETag != nil {
-		etag = *req.ETag
+	marshalledEntity, err := r.marshal(req)
+	if err != nil {
+		return err
 	}
-	entity.OdataEtag = etag
 
 	// InsertOrReplace does not support ETag concurrency, therefore we will use Insert to check for key existence
 	// and then use Update to update the key if it exists with the specified ETag
+	_, err = r.client.AddEntity(ctx.Background(), marshalledEntity, nil)
 
-	err := entity.Insert(storage.FullMetadata, nil)
 	if err != nil {
 		// If Insert failed because item already exists, try to Update instead per Upsert semantics
 		if isEntityAlreadyExistsError(err) {
```

Comment:
Can you short-circuit this?

```go
if err == nil {
    return nil
} else if !isEntityAlreadyExistsError(err) {
    // Any other unexpected error on Insert is propagated to the caller
	return err
}

// Rest of code

Reply:
No it really needs to be this very exact code. I know it's weird... basically, no error is code.. and a specific error is also good.. but all other errors are bad :p

Reply:
Yeah that's why I was saying to format it that way. Just so there's less nesting of `if`s. But not a blocker :)